src/format/Quoting.cc

```diff
@@ -122,3 +123,29 @@ Format::QuoteMimeBlob(const char *header)
     return buf;
 }
 
+SBuf
+Format::DquoteString(const SBuf &src)
+{
+    static CharacterSet unreserved = CharacterSet("unreserved", "\"\\").complement();
+
+    SBuf quotedStr;
+    quotedStr.append('"');
+
+    Parser::Tokenizer tok(src);
+    while (!tok.atEnd()) {
+        SBuf found;
+        if (tok.prefix(found, unreserved))
+            quotedStr.append(found);
+
+        if (!tok.atEnd()) {
+            char c = tok.buf()[0];
+            quotedStr.append('\\');
+            quotedStr.append(c);
+            tok.skip(c);
+        }
```

Comment:
Sure. Out of scope for this PR though.

Reply:
I disagree. There is a way to implement this logic without improving Tokenizer and without introducing problems flagged in this change request. The suggested improvement is just the best way forward IMO (and can be done here or in a separate PR).

Reply:
There are many ways to implement this. I have chosen the one used for minimal complexity given the APIs currently in Squid.

Reply:
Opened PR #1545 for this

Reply:
> I have chosen the one used for minimal complexity given the APIs currently in Squid.

We should not add risky code (and then copy it all over the place) when it is easy to add safe one. If we had a Tokenizer fairy that would quickly create the missing methods for us _and_ cleanup after our rushed PRs, OR if adding a missing Tokenizer method were difficult, it would be OK to cut this corner, but that fairy does not exist. We have to do the legwork required to move the code forward rather than fix one problem while adding two new ones.

Recent experience reminded us how difficult it is to write correct parsing code -- we spend three or four review iterations fixing fresh PR bugs because we cut a few corners earlier. We should insist on principles that make repeated bugs less likely. Avoiding intermediate EOF checks and avoiding raw buffer accesses are examples of such principles.